cylc/flow/task_queue.py

```diff
@@ -0,0 +1,197 @@
+            if qname == "type":
+                continue
+            if qname == self.Q_DEFAULT:
+                # Add all tasks to the default queue.
+                qmembers = set(all_task_names)
+            else:
+                qmembers = set()
+                for mem in queue["members"]:
+                    if mem in descendants:
+                        # Family name.
+                        for fmem in descendants[mem]:
+                            if (fmem not in descendants
+                                    and fmem in all_task_names):
+                                # Task name.
+                                qmembers.add(fmem)
+                    elif mem in all_task_names:
+                        # Task name.
+                        qmembers.add(mem)
+            queues[qname] = {}
+            queues[qname]["members"] = qmembers
+            queues[qname]["limit"] = queue["limit"]
+        return queues
+
+    def _make_indep(self, in_queues: dict) -> dict:
+        """For classic queues tasks cannot belong to multiple limit groups.
+
+        Latest assigned group takes precedence. So, for example, the "default"
+        group will only contain tasks not assigned to any other queue.
+
+        """
+        queues: Dict[str, Any] = {}
+        seen: Dict[str, str] = {}
+        for qname, qconfig in in_queues.items():
+            queues[qname] = {}
+            queues[qname]["members"] = qconfig["members"]
+            queues[qname]["limit"] = qconfig["limit"]
+            if qname == self.Q_DEFAULT:
+                continue
+            for qmem in qconfig["members"]:
+                # Remove from default queue
+                try:
+                    queues[self.Q_DEFAULT]["members"].remove(qmem)
+                except KeyError:
+                    # Already removed.
+                    pass
+                if qmem in seen:
+                    # Override previous queue assignment.
+                    oldq = seen[qmem]
+                    queues[oldq]["members"].remove(qmem)
+                else:
+                    queues[qname]["members"].add(qmem)
+                seen[qmem] = qname
+        return queues
+
+    def _configure_limiters(self, queues: dict) -> None:
+        """Create a Limiter for each limit group."""
+        for name, config in queues.items():
+            self.limiters[name] = Limiter(config["limit"], config["members"])
+
+    def add(self, itask: TaskProxy) -> None:
+        """Queue a task."""
+        LOG.debug(f"Queue add: {itask.identity}")
+        itask.state.reset(is_queued=True)
+        itask.reset_manual_trigger()
+        self.task_deque.appendleft(itask)
+
+    def release(self, active: Counter[str]) -> List[TaskProxy]:
+        """Release queued tasks."""
+        released: List[TaskProxy] = []
+        rejects: List[TaskProxy] = []
+        while True:
+            try:
+                candidate = self.task_deque.pop()
```

Comment:
For large flows with small limits the work of emptying and rebuilding the dequeue could become quite expensive.

Could we just iterate over a list and only remove tasks when dequeued? That would only perform data structure operations when things change:

```python
for candidate in self.task_list:
     if candiate.state.is_held or any(limiter...):
          # no need to track rejects or rebuild the dequeue afterwards
          continue
     self.task_list.remove(candidate)
     released.append(candidate)
     ...
```

Reply:
I've converted to a list because it's simpler. There are some minuses though: I have to iterate over a copy to allow item removal; and removing list items is more expensive than removing deque items.